prov/sockets/src/sock_ep_msg.c

```diff
@@ -614,74 +634,62 @@ static int sock_ep_cm_connect(struct fid_ep *ep, const void *addr,
 	memcpy(&req->src_addr, _ep->attr->src_addr, sizeof(req->src_addr));
 	memcpy(&handle->dest_addr, addr, sizeof(handle->dest_addr));
 
+	_ep->attr->info.handle = (void*) handle;
 	handle->ep = _ep;
 	handle->req = req;
+	handle->cm_head = &_ep->attr->domain->cm_head;
+	pthread_mutex_init(&handle->mutex_cond.mutex, NULL);
+	pthread_cond_init(&handle->mutex_cond.cond, NULL);
 	if (paramlen) {
 		handle->paramlen = paramlen;
 		memcpy(handle->cm_data, param, paramlen);
 	}
 
-	if (_ep->attr->cm.listener_thread &&
-	    pthread_join(_ep->attr->cm.listener_thread, NULL))
-		SOCK_LOG_DBG("failed to join cm listener\n");
-
-	if (pthread_create(&_ep->attr->cm.listener_thread, NULL,
-			   sock_ep_cm_connect_handler, handle)) {
-		SOCK_LOG_ERROR("failed to create cm thread\n");
+	sock_fd = ofi_socket(AF_INET, SOCK_STREAM, 0);
+	if (sock_fd < 0) {
+		SOCK_LOG_ERROR("no socket\n");
 		goto out;
 	}
+
+	ofi_straddr_dbg(&sock_prov, FI_LOG_EP_CTRL, "Connecting to address",
+			&handle->dest_addr);
+	sock_set_sockopts_conn(sock_fd);
+	ret = connect(sock_fd, (struct sockaddr *)&handle->dest_addr,
+		      sizeof(handle->dest_addr));
+	if (ret < 0) {
+		SOCK_LOG_ERROR("connect failed : %s\n",
+			       strerror(ofi_sockerr()));
+		goto err;
+	}
+
+	if (sock_cm_send(sock_fd, req, sizeof(*req)))
```

Comment:
The socket must be set to non-blocking.  Connect cannot guarantee to complete immediately.  We need async handling here prior to attempting to send on the socket.

Reply:
Thanks for the insightful comment. I have a separate patch that allows an asynchronous handling of the connection requests. While reading the socket provider, I've figured out that there is a similar issue with the function `sock_ep_connect` in `sock_conn.c` where `connect()` might block. We should probably also revisit this code (see https://github.com/ofiwg/libfabric/issues/2586).